### DISCUSSION TITLE
How does reset_all work? and Virtual input example

### DISCUSSION THREAD

**Discussion by ndgipper** [fasteddy516/CircuitPython_JoystickXL#33] [Q&A]
I'm trying to follow the reset_all path.  It looks like it calls update() which reads the devices current values and thus an IDLE report never gets sent.  Am I reading this wrong?  I expected reset_all to send a truly idle report.

The reason I'm on this path is I'm trying to find a way to call update_axis,buttons,hats() and then replace an axis value to be the same as a different axis value before the report is sent.   I want to do this when a certain button is True.  Otherwise, each axis would get its correct value reported.

However, I've gotten distracted trying to see how reset_all works.  I'm probably missing something.  Apologies for asking a dumb question.

**Comment by fasteddy516** [comment]
Hi @ndgipper,

You've got me stumped as far as `reset_all()` goes.  You're absolutely correct - the internal `_*_states` lists all get set to default/idle values, but then get overwritten with the real-time values when `update()` gets called.  It's been quite awhile since I originally developed JoystickXL and I can't honestly remember what `reset_all` was supposed to accomplish.  I have a feeling it was in place before the `update()` mechanism was fully implemented and has been lurking in the background doing nothing useful ever since.

Definitely not a dumb question, so no need to apologize!

**Reply by ndgipper** [reply]
Thanks.
I've managed to get what I wanted quite easily.  Thanks for writing this handy library.
I'm using a TCA8418 and ads7830 with an M4feather to get 64 buttons and 8 axes.
The .source_value hook is quite clever.

**Comment by fasteddy516** [comment]
As far as your use case is concerned, you should be able to accomplish that without having to call the individual `update_*` methods at all.  The (admittedly not properly documented) trick is to process the raw axis values on your own rather than tying JoystickXL directly to pins on your board.  The example below creates a joystick with 1 button and 1 axis, and feeds one of two analog input values to the axis depending on the button state:

```python
import board
from analogio import AnalogIn
from joystick_xl.inputs import Axis, Button
from joystick_xl.joystick import Joystick

joystick = Joystick()

joystick.add_input(
    Button(board.GP0),  # this button is tied to pin GP0 on my CircuitPython board (a Raspberry Pi Pico in this case)
    Axis(),  # axis[0] is the x-axis - notice that I haven't tied it directly to a pin on the board
)

raw_axis_0 = AnalogIn(board.A0)
raw_axis_1 = AnalogIn(board.A1)

joystick.update()

while True:
    if joystick.button[0].is_pressed:
        joystick.axis[Axis.X].source_value = raw_axis_0.value
    else:
        joystick.axis[Axis.X].source_value = raw_axis_1.value
    joystick.update()
```

If you have code you're willing to share - or just more information about your use case - I should be able to help you sort through it.

**Reply by ndgipper** [reply]
Awesome!  That looks exactly like what I just came up with!
You could put that code in an example for future reference.

**Comment by ndgipper** [comment]
Here's the work in progress.  I only have 2 axes active right now.
Feel free to use this in an example if you'd like.
The idea was to have a switch on this USB throttle that I'm building that will have two throttle pots L/R but most of the time the R throttle will send its value as both L and R.  But, sometimes they would be split - when the switch is the other way.

```
import board
import time
import digitalio
import analogio
from joystick_xl.inputs import Axis, Button, Hat
from joystick_xl.joystick import Joystick

# get the library for the ADS7830 ADC breakout board
import adafruit_ads7830.ads7830 as ADC
from adafruit_ads7830.analog_in import AnalogIn

# get the library for the TCA8418 ADC breakout board
from adafruit_tca8418 import TCA8418

# create the I2C support for the ads7830 board and the tca8418 board
i2c = board.I2C()

# set up the button matrix 8x8
tca = TCA8418(i2c)
KEYPADPINS = (TCA8418.R0, TCA8418.R1, TCA8418.R2, TCA8418.R3,
              TCA8418.R4, TCA8418.R5, TCA8418.R6, TCA8418.R7,
              TCA8418.C0, TCA8418.C1, TCA8418.C2, TCA8418.C3,
              TCA8418.C4, TCA8418.C5, TCA8418.C6, TCA8418.C7,)
for pin in KEYPADPINS:
    tca.keypad_mode[pin] = True      # make it a keypad pin instead of gpio pin
    tca.event_mode_fifo[pin] = True  # event will go into FIFO queue
tca.key_intenable = True

# set up the led in case we want it
led = digitalio.DigitalInOut(board.D13)
led.switch_to_output()

# Set up a digital input for the throttle function switch.
# This will not be part of the usb report.
sw0 = digitalio.DigitalInOut(board.D5)
sw0.switch_to_input(pull = digitalio.Pull.DOWN)

# instantiate a Joystick object
js = Joystick()


# set up the axes using the ads7830 board (adc)
adc = ADC.ADS7830(i2c)
pot0 = AnalogIn(adc,0)
pot1 = AnalogIn(adc,1)
#pot2 = AnalogIn(adc,2)
#pot3 = AnalogIn(adc,3)
#pot4 = AnalogIn(adc,4)
#pot5 = AnalogIn(adc,5)
#pot6 = AnalogIn(adc,6)
#pot7 = AnalogIn(adc,7)

# activate inputs in the joystick object (js)
for _ in range(64):
    js.add_input( Button(active_low=False))

js.add_input(
   Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 0
   Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 1
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 2
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 3
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 4
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 5
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 6
   #Axis(),  # a virtual Axis that we'll fill in during the loop  Axis 7
)

print('loop starting')

while True:
    # Axis 0 is virtual so we need to set the value.
    js.axis[0].source_value = pot0.value
    # Axis 1 is virtual so we need to set the value.
    # If sw0 is True we'll use the value from Axis0.
    # Otherwise, we'll use the value from pot1 for Axis1.
    if sw0.value:
        js.axis[1].source_value = js.axis[0].source_value
    else:
        js.axis[1].source_value = pot1.value
    # Remaining virtual axes are like pot0
    #js.axis[2].source_value = pot2.value
    #js.axis[3].source_value = pot3.value
    #js.axis[4].source_value = pot4.value
    #js.axis[5].source_value = pot5.value
    #js.axis[6].source_value = pot6.value
    #js.axis[7].source_value = pot7.value

    # Update the virtual buttons if something was triggered by the tca.
    if tca.key_int:
        nevts = tca.events_count
        for _ in range(nevts):
            keyevent = tca.next_event
            ib = (keyevent & 0x7F) - 1
            val = bool(keyevent & 0x80)
            #print(keyevent,ib,val)
            js.button[ib].source_value = val
        tca.key_int = True  # clear the IRQ

    # Non-virtual buttons/axes get updated from the board.Dx pins
    # and any board.Ax pins during the update() function.
    js.update()

    # delay before looping
    time.sleep(0.10)

```

**Reply by ndgipper** [reply]
I've edited the code above to remove an erroneous comment.

**Comment by ndgipper** [comment]
My code is more "readable" to me the way I've made all of the Axis inputs virtual.  However, I think it's only necessary that the Axis 1 (pot1) input be virtual.  Perhaps a slicker example would should the mixture of virtual and non-virtual axes.
